pkg/agent/controller/egress/ip_assigner_linux.go

```diff
@@ -0,0 +1,186 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package egress
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/vishvananda/netlink"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/util"
+	"antrea.io/antrea/pkg/agent/util/arping"
+)
+
+var ipv6NotSupportErr = errors.New("IPv6 not supported")
+
+type ipAssigner struct {
+	egressInterface *net.Interface
+	egressLink      netlink.Link
+	egressRunDir    string
+	mutex           sync.RWMutex
+	assignedEgress  map[string]string
+}
+
+// NewIPAssigner return an *ipAssigner.
+func NewIPAssigner(nodeIPAddr net.IP, dir string) (*ipAssigner, error) {
+	_, egressInterface, err := util.GetIPNetDeviceFromIP(nodeIPAddr)
+	if err != nil {
+		return nil, fmt.Errorf("get IPNetDevice from ip %v error: %+v", nodeIPAddr, err)
+	}
+	ifaceName := egressInterface.Name
+	egressLink, err := netlink.LinkByName(ifaceName)
+	if err != nil {
+		return nil, fmt.Errorf("get netlink by name(%s) error: %+v", ifaceName, err)
+	}
+	a := &ipAssigner{
+		egressInterface: egressInterface,
+		egressLink:      egressLink,
+		egressRunDir:    dir,
+		assignedEgress:  make(map[string]string),
+	}
+	if err := a.initAssignedIPFiles(); err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+func (a *ipAssigner) initAssignedIPFiles() (err error) {
+	files, err := ioutil.ReadDir(a.egressRunDir)
+	if err != nil {
+		if err := os.MkdirAll(a.egressRunDir, os.ModeDir); err != nil {
+			return fmt.Errorf("error when creating egress run dir: %v", err)
+		}
+		return nil
+	}
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	for _, file := range files {
+		egressName := file.Name()
+		fileName := ipSavedFile(a.egressRunDir, egressName)
+		if ip, err := ioutil.ReadFile(fileName); err == nil {
+			a.assignedEgress[egressName] = string(ip)
+		}
+	}
+	return
+}
+
+// AssignEgressIP assign an egressIP and save persistent files.
+func (a *ipAssigner) AssignEgressIP(egressIP, egressName string) error {
+	egressSpecIP := net.ParseIP(egressIP)
+	if egressSpecIP == nil {
+		return fmt.Errorf("invalid IP %s", egressIP)
+	}
+	addr := netlink.Addr{IPNet: &net.IPNet{IP: egressSpecIP, Mask: net.CIDRMask(32, 32)}}
+	ifaceName := a.egressInterface.Name
+	if err := netlink.AddrAdd(a.egressLink, &addr); err != nil {
+		return fmt.Errorf("failed to add ip %v to interface %s: %v", addr, ifaceName, err)
+	}
+	if err := a.saveIPAssignFile(egressIP, egressName); err != nil {
+		return fmt.Errorf("failed to save egress ip assign, egress %s, ip: %s: %v", egressName, egressIP, err)
+	}
+	isIPv4 := egressSpecIP.To4()
+	if isIPv4 != nil {
+		if err := arping.GratuitousARPOverIface(isIPv4, a.egressInterface); err != nil {
+			return fmt.Errorf("Failed to send gratuitous ARP: %v", err)
```

Comment:
nit: should not be capitalized, same for the one below

Reply:
done